klee-replay: make it safe to run multiple instances in parallel and other improvements

ccadar opened the pull request:
This fixes the Gen*Bout.c failures that we've experienced recently on Travis CI

ccadar commented:
@MartinNowack @251 this one is ready to review, it addresses the intermittent failures in Gen*Bout.c we've been discussing

251 commented on tools/klee-replay/file-creator.c:
```diff
@@ -149,7 +150,7 @@ static int create_char_dev(const char *fname, exe_disk_file_t *dfile,
 
     if (symlink(name, fname) == -1) {
-      fprintf(stderr, "unable to create sym link to tty\n");
+      fprintf(stderr, "KLEE-REPLAY: ERROR: unable to create sym link to tty\n");
```
All `fprintf`s without format string should be `fputs`.

ccadar replied:
Why?

251 replied:
fprintf is way more complex and burns CPU cycles for nothing.

ccadar replied:
OK, but the difference doesn't matter in most cases (such as here, for error reporting) and nowadays many compilers do this type of optimization themselves.  The reason I prefer `(f)printf` slightly is that I can then use everywhere, regardless of whether it's a simple string or not.

251 replied:
I don't feel strongly about this, I only think it's a bad habit to rely on a compiler optimisation. Why use a function with full-blown var-arg handling and string parsing for sth. that simple in the first place? Also it makes reading the code harder as one automatically tries to find format specifiers or arguments... but it's just my opinion and there are more important issues than ~20 misused `fprintf`s.

ccadar replied:
OK, thanks for being flexible ;)   I could agree with this for an isolated instance, but in this case, we have a ton of `fprintf`s in there, and I find it less readable to switch from `fprintf`to `fputs` or vice-versa.  For instance:
```
fprintf(stderr, "Usage: %s [option]... <executable> <ktest-file>...\n", progname);
fprintf(stderr, "   or: %s --create-files-only <ktest-file>\n", progname);
fprintf(stderr, "\n");
fprintf(stderr, "-r, --chroot-to-dir=DIR  use chroot jail, requires CAP_SYS_CHROOT\n");
```
the last two `fprintf`s could be replaced by `fputs`, but wouldn't you find it a bit ugly to do so?

251 replied:
No, because here I would just use a single `fprintf` as I did in [gen-bout.c](https://github.com/klee/klee/blob/64508909bcbd52d6a06ec0f299b3752368d10e31/tools/gen-bout/gen-bout.cpp#L50). (The formatting is still weird, I kept the original one.)

ccadar replied:
I see :)  But isn't it slower than mixing `fputs` and `fprintf`? ;) 
In any case, feel free to submit a patch if you want to do a pass over the way messages are printed in there.  A better option would be perhaps to offer functions such as `klee_replay_error()` etc. as we do in the core KLEE tool.

ccadar commented on tools/klee-replay/file-creator.c:
```diff
@@ -375,12 +377,11 @@ static void create_file(int target_fd,
-    target = tmpname;
-  }
+  if (target_name)
+    sprintf(tmpname, "%s/%s", tmpdir, target_name);
```
OK

ccadar replied:
Done

ccadar commented on test/Runtime/POSIX/Replay.c:
```diff
@@ -1,10 +1,7 @@
-
-// RUN: FileCheck --input-file=A --check-prefix=CREATE_FILES_ONLY %s
-// CREATE_FILES_ONLY: abc
+// RUN: %klee-replay --create-files-only %t.klee-out/test000001.ktest 2>%t.txt
```
Good catch :)  Since now all files are created in a temporary directory, `--input-file=A` doesn't work anymore.  If we want to keep this check, we would either need to parse the output of `klee-replay` for the name of the temporary directory, or add a feature to `klee-replay` to let users specify the name of the directory.  But I'm not sure that feature would be useful outside this check.

MartinNowack replied:
In that case I'm for removing this check. The remaining checks are sufficient. Please, just remove the `2>%t.txt` at the end.

ccadar replied:
Done

ccadar commented:
@MartinNowack @251 Thanks for the feedback!  
@MartinNowack I replaced `sprintf` to `snprintf` throughout the codebase with a separate commit
@251 I decided not to replace `fprintf` by `fputs` in those few cases, as the code looks more readable with `fprintf` everywhere (why were these functions designed to take the string and the stream in a different order and have different newline behaviour!?) and as I argued above, performance is not really an issue here.  But if you feel very strongly about this, let me know.

Finally, I removed a bunch of unneeded code.

ccadar commented on tools/klee-replay/file-creator.c:
```diff
@@ -507,7 +509,9 @@ static void check_file(int index, exe_disk_file_t *dfile) {
     name[0] = 'A' + index;
     name[1] = '\0';
-    res = stat(name, &s);
+    sprintf(fullname, "%s/%s", replay_dir, name);
```
Done

ccadar commented:
I also moved the Gen*Bout tests outside the `test/Runtime/POSIX` directory as they don't need POSIX support.   For instance, these tests will also run on macOS now.

ccadar commented:
@MartinNowack @251 this is ready for another look

MartinNowack commented:
@ccadar Thanks a lot for the PR and especially all the effort to find the issue in the first place.

kren1 commented on tools/klee-replay/klee-replay.c:
```diff
@@ -192,12 +192,17 @@ static void run_monitored(char *executable, int argc, char **argv) {
 #endif
 
     if (!rootdir) {
+      if (chdir(replay_dir) != 0) {
```
This is broken now, because executable can no longer be a relative path to the current dir. For example

`klee-replay b2sum path-to-ktest-file` no longer works because `execv` can't find b2sum in the `replay_dir`
